stan/math/rev/fun/tcrossprod.hpp

```diff
@@ -21,41 +21,26 @@ namespace math {
  * @param M Matrix to multiply.
  * @return M times its transpose.
  */
-template <typename T, require_eigen_vt<is_var, T>* = nullptr>
-inline Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime>
-tcrossprod(const T& M) {
-  if (M.rows() == 0) {
-    return {};
-  }
-  // if (M.rows() == 1)
-  //   return M * M.transpose();
-
-  // WAS JUST THIS
-  // matrix_v result(M.rows(), M.rows());
-  // return result.setZero().selfadjointView<Eigen::Upper>().rankUpdate(M);
-
-  Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime> MMt(M.rows(),
-                                                                     M.rows());
-
-  vari** vs
-      = reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
-          (M.rows() * M.cols()) * sizeof(vari*)));
-  int pos = 0;
-  for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < M.cols(); ++n) {
-      vs[pos++] = M(m, n).vi_;
-    }
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    MMt.coeffRef(m, m)
-        = var(new internal::dot_self_vari(vs + m * M.cols(), M.cols()));
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < m; ++n) {
-      MMt.coeffRef(n, m) = MMt.coeffRef(m, n) = dot_product(M.row(m), M.row(n));
-    }
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline plain_type_t<T> tcrossprod(const T& M) {
+  using T_double = Eigen::MatrixXd;
+  using T_var = plain_type_t<T>;
+
+  if (M.size() == 0) {
+    return M;
   }
-  return MMt;
+
+  arena_t<plain_type_t<T>> arena_M = M;
+  arena_t<Eigen::MatrixXd> arena_M_val = value_of(arena_M);
```

Comment:
Hmm, I think we need to do something here to avoid the copy in the varmat case. The easiest option would be to change the var<matrix> to use the arena matrix internally instead of maps, so the arena_matrix constructor here will be able to do a shallow copy.

Reply:
^ I was thinking something similar yesterday. idt it would add any unreasonable bulk and would make stuff flow a lot more naturally. I want to get in all the big changes from #2064 then we can do this in a separate PR

Reply:
Actually this was pretty easy-peasy!

https://github.com/stan-dev/math/compare/feature/multi-expression-vari-views...feature/vari_value-arena-matrix

Reply:
Nice, can you open a separate pull request with this? I don't wanna slow this up going through here.

Also the code here: https://github.com/stan-dev/math/pull/2106#discussion_r497788511 reminds me that maybe this would work:
```
const auto& arena_M_val = to_arena(value_of(arena_M));
```

If it see's an arena variable it just forwards it along I think.